### PR TITLE
Support async cancellations

### PR DESCRIPTION
**Pull request by karosis88**
# Problem

Cancellations in the middle of a request/response cycle are not supported by the httpcore.

Something like this should brake the connection pool

```python
import anyio
import httpcore

async def main():
    pool = httpcore.AsyncConnectionPool()
    with anyio.move_on_after(1):
        await pool.request(
            "GET", "http://example.com"
        )  # long running request
    print(pool.connections)  # One active connection that will remain active indefinitely
    # [<AsyncHTTPConnection ['http://example.com:80', HTTP/1.1, ACTIVE, Request Count: 1]>]

anyio.run(main)
```

# TODO

- [x] Write a failing test
- [ ] Make a test pass

### CODE REVIEW

File: tests/test_cancecllations.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@pytest.mark.anyio
+async def test_async_cancellation():
+    pool = httpcore.AsyncConnectionPool(network_backend=SlowBackend())
+    with anyio.move_on_after(1):
+        await pool.request(
+            "GET", "http://example.com"
+        )
+    assert not pool.connections
```
Comment (tomchristie) [inline tests/test_cancecllations.py:42]: I don't think this is quite correct.
The behaviour we'd want to see here is that there *is* a connection in the pool, but it is IDLE.

File: tests/test_cancecllations.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        timeout: Optional[float] = None,
+        local_address: Optional[str] = None,
+        socket_options: typing.Optional[typing.Iterable[SOCKET_OPTION]] = None,
+    ) -> AsyncNetworkStream:
+        return SlowStream()
+
+
+@pytest.mark.anyio
+async def test_async_cancellation():
+    pool = httpcore.AsyncConnectionPool(network_backend=SlowBackend())
```
Comment (tomchristie) [inline tests/test_cancecllations.py:37]: Let's go with the neater usage of `async with ... as pool:`.
That's what we use elsewhere in the tests.

File: tests/test_cancecllations.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    with anyio.move_on_after(1):
```
Comment (tomchristie) [inline tests/test_cancecllations.py:38]: A nice quick timeout in our test cases would be good.
Is 0.001 okay here?

File: tests/test_cancecllations.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import typing
+from typing import Optional
+
+import anyio
+import pytest
+
+import httpcore
+from httpcore.backends.auto import AutoBackend
+from httpcore.backends.base import SOCKET_OPTION, AsyncNetworkBackend
+from httpcore.backends.mock import AsyncNetworkStream
```
Comment (tomchristie) [inline tests/test_cancecllations.py:10]: Could we just use a top-level `import httpcore` instead?

File: tests/test_cancecllations.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import typing
```
Comment (Pliner) [inline tests/test_cancecllations.py:1]: There is a typo in the filename (test_cancecllations -> test_cancellations).

File: tests/test_cancecllations.py (hunk 1 above)
Comment (karosis88) [reply tests/test_cancecllations.py]: I thought about it and can't remember when, but I remember thinking that this is much safer.

**Comment by karosis88** [comment]
First, we must decide how we will support "CancelScope" while keeping sync and async interfaces similar. I'm not sure whether it's better to do it in another PR or in this one.

**Comment by tomchristie** [comment]
> First, we must decide how we will support "CancelScope" while keeping sync and async interfaces similar.

Okay. Before we get to implementation, could you start off by showing me *where* we need shielding rather than *how* we're going to implement it?

In order to have confidence that we're *really getting it right* I'd be interested in looking at that from the ground up...

Suppose we're just working directly with an `HTTPConnection` instance directly, rather than with a connection pool. Where would we need cancellation shielding?

**Comment by karosis88** [comment]
https://github.com/encode/httpcore/blob/eb5957d60b240a0deec503a4acca0e69da2833db/httpcore/_async/http11.py#L220-L232

This function appears to be the main source of the problem; the issue here is that if we call this method while cancelled, the resources will not be closed.

**Comment by tomchristie** [comment]
Perhaps we should add a test case at that level of the API?

**Comment by karosis88** [comment]
Sounds good

File: tests/test_cancellations.py [hunk 6]
```diff
@@ -35,3 +38,39 @@ async def test_async_cancellation():
...
+    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
+        # mock
+        conn._state = HTTP11ConnectionState.ACTIVE
+
+        with anyio.CancelScope() as cancel_scope:
+            async with conn._state_lock:
+                cancel_scope.cancel()
+                await conn._response_closed()
+        assert conn._state != HTTP11ConnectionState.ACTIVE
+        assert conn._state in (HTTP11ConnectionState.IDLE, HTTP11ConnectionState.CLOSED)
```
Comment (tomchristie) [inline tests/test_cancellations.py:56]: Okay so, challenge...

Tests ought to be against our public API. In this case, using `conn.request(...)` with a cancellation.

File: tests/test_cancellations.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+@pytest.mark.anyio
+async def test_h11_response_closed():
+    origin = httpcore.Origin(b"https", b"example.com", 443)
+    stream = httpcore.AsyncMockStream([])
+    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
+        # mock
+        with anyio.CancelScope() as cancel_scope:
+            cancel_scope.cancel()
+            await conn.request("GET", "https://example.com")
+        assert conn._state == HTTP11ConnectionState.CLOSED
```
Comment (tomchristie) [inline tests/test_cancellations.py:52]: I assume we want the test to mirror the same style as the `test_async_cancellation` case above.

Something like this?...

```suggestion
    stream = SlowStream()
    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
        with anyio.move_on_after(0.001):
            await pool.request("GET", "http://example.com")
        assert conn.is_closed()
```

Comment (karosis88) [reply tests/test_cancellations.py]: This looks much better, thanks

File: tests/test_cancellations.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import pytest
+
+import httpcore
+
+
+class SlowStream(httpcore.AsyncNetworkStream):
+    async def write(
+        self, buffer: bytes, timeout: typing.Optional[float] = None
+    ) -> None:
+        await httpcore._backends.auto.AutoBackend().sleep(2)
```
Comment (tomchristie) [inline tests/test_cancellations.py:14]:
```suggestion
        await anyio.sleep(2)
```

File: tests/test_cancellations.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import typing
+from typing import Optional
```
Comment (tomchristie) [inline tests/test_cancellations.py:2]: Let's just use the top level `import typing` style throughout.

File: tests/test_cancellations.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+
+@pytest.mark.anyio
+async def test_h2_response_closed():
+    origin = httpcore.Origin(b"http", b"example.com", 80)
+    stream = SlowStream()
+    async with httpcore.AsyncHTTP2Connection(origin, stream) as conn:
+        with anyio.move_on_after(0.001):
+            await conn.request("GET", "http://example.com")
+        assert conn.is_closed()
```
Comment (tomchristie) [inline tests/test_cancellations.py:57]: Okay these are looking great - good starting point for us working through this really comprehensively.

Comment (karosis88) [reply tests/test_cancellations.py:56]: If we're done with failing tests, I believe we should decide how and where we want to see shield logic.

**Comment by tomchristie** [comment]
Starting with the HTTP/1.1 case it looks to me like we want...

* Two test cases - the existing one that we currently have for a timeout during writing, and an additional one for a timeout during reading.
* Shielding semantics around our exception handling in the `handle_async_request` method...

```python
        except BaseException as exc:
            with ShieldFromCancellation():
                async with Trace("response_closed", logger, request) as trace:
                    await self._response_closed()
            raise exc
```

* Shielding semantics around our exception handling in the byte stream `__aiter__` method...

```python
        except BaseException as exc:
            # If we get an exception while streaming the response,
            # we want to close the response (and possibly the connection)
            # before raising that exception.
            with ShieldFromCancellation():
                await self.aclose()
            raise exc
```

I'm not presuming how we implement the `ShieldFromCancellation` here, just presenting what I think the semantics ought to look like. (Based on `trio`'s design in this problem space. Because *structured concurrency FTW*.)

**Comment by karosis88** [comment]
To me, it appears to be completely correct.

**Comment by karosis88** [comment]
> Two test cases - the existing one that we currently have for a timeout during writing, and an additional one for a timeout during reading.

Are you referring to slow-write test for `ConnectionPool` or `HTTP11Connection`

**Comment by tomchristie** [comment]
> Are you referring to slow-write test for ConnectionPool or HTTP11Connection?

I'm suggesting that we'll 


```python
class SlowReadStream(httpcore.AsyncNetworkStream):
    def __init__(self, buffer):
        self.buffer = buffer

    async def write(self, buffer, timeout = None):
        pass

    async def read(self, max_bytes: int, timeout: typing.Optional[float] = None) -> bytes:
        if not self._buffer:
            await trio.sleep(1)
        else:
            return self._buffer.pop(0)

    async def aclose(self):
        pass

stream = SlowReadStream([
    b"HTTP/1.1 200 OK\r\n",
    b"Content-Type: plain/text\r\n",
    b"Content-Length: 1000\r\n",
    b"\r\n",
    b"Hello, world!...", 
    # The stream will hang after this portion of the incomplete response,
    # so that we're able to test our async cancellation semantics.
])
```

**Comment by karosis88** [comment]
Is a buffer really necessary?

**Comment by tomchristie** [comment]
> Is a buffer really necessary?

You'd need this or something *like* this in order to test the cancellation handling for `HTTP11ConnectionByteStream.__aiter__`.

**Comment by tomchristie** [comment]
Okay, want to try...

* Adding the `ShieldCancellation` from https://github.com/encode/httpcore/pull/726
* Making the first HTTP/1.1 test pass.
* Making the second HTTP/1.1 test pass.

File: tests/test_cancellations.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+            b"Content-Length: 1000\r\n",
+            b"\r\n",
+            b"Hello, world!...",
+        ]
+    )
+    async with httpcore.AsyncHTTP11Connection(origin, stream) as conn:
+        with anyio.move_on_after(0.001):
+            async with conn.stream("GET", "http://example.com") as resp:
+                async for chunk in resp.aiter_stream():
+                    pass
```
Comment (tomchristie) [inline tests/test_cancellations.py:94]: Just `conn.request(...)` would be sufficient here.

(We'll still end up calling into `HTTP11ConnectionByteStream.__aiter__`)

**Comment by tomchristie** [comment]
Thanks so much for your work on this.
I've started pulling it into #726, so I think we can close this one as being superseeded at this point.

**Pull request closed without merging**